data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/client/DataService.java

```diff
@@ -0,0 +1,82 @@
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.CreateEnvironmentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.CreateEnvironmentResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.DeleteEnvironmentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.DeleteEnvironmentResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.DescribeEnvironmentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.DescribeEnvironmentResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.DescribeEnvironmentRevisionResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.ListClustersRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.ListClustersResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.ListEnvironmentRevisionsRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.ListEnvironmentRevisionsResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.ListEnvironmentsRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.ListEnvironmentsResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.StartDeploymentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.StartDeploymentResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.UpdateEnvironmentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.UpdateEnvironmentResponse;
+
+public interface DataService {
+
+  /** Creates an environment record and a new revision record. */
+  CreateEnvironmentResponse createEnvironment(CreateEnvironmentRequest request)
+      throws EnvironmentExistsException, InvalidParameterException, ServiceException;
+
+  /** Creates a new environment revision for the environment. */
+  UpdateEnvironmentResponse updateEnvironment(UpdateEnvironmentRequest request)
+      throws EnvironmentNotFoundException, InvalidParameterException, ServiceException;
+
+  /** Returns the environment record. */
+  DescribeEnvironmentResponse describeEnvironment(DescribeEnvironmentRequest request)
+      throws EnvironmentNotFoundException, InvalidParameterException, ServiceException;
+
+  /** Lists all environments with a filter. */
+  ListEnvironmentsResponse listEnvironments(ListEnvironmentsRequest request)
+      throws InvalidParameterException, ServiceException;
+
+  /** Deletes the provided environment if inactive or if forceDelete is true */
+  DeleteEnvironmentResponse deleteEnvironment(DeleteEnvironmentRequest request)
+      throws EnvironmentNotFoundException, EnvironmentActiveException, InvalidParameterException,
+          ServiceException;
+
+  /** Returns the requested environment revision. */
+  DescribeEnvironmentRevisionResponse describeEnvironmentRevision(
+      DescribeEnvironmentRequest request)
+      throws EnvironmentNotFoundException, EnvironmentRevisionNotFoundException,
```

Comment:
I think it might be worth avoiding resource-specific subclasses of NotFoundException. Perhaps model these as `ResourceNotFoundException` with `resourceType` and `resourceId` fields?

Reply:
I think it's useful to have the specific exceptions on the dataservice to aid with debugging. They can be mapped to resourcenotfound on the FE

Reply:
I guess that would make it hard to populate which resource on the frontend :/

Reply:
Ayup :/